File: accountant.py
import time
import logging
from typing import Dict, Any

class PnLAccountant:
    """
    [OXX TACTICAL ACCOUNTANT]
    A physically decoupled math engine for tracking session PnL and 
    calculating pre-flight trade hurdles.
    """
    def __init__(self, taker_fee_rate: float = 0.0035, maker_fee_rate: float = 0.0020):
        # We default to standard VIP 0 rates if the API hasn't updated yet
        self.taker_rate = abs(float(taker_fee_rate))
        self.maker_rate = abs(float(maker_fee_rate))
        self.tier_label = "PENDING"
        self.realized_pnl_gross = 0.0
        self.total_fees_paid = 0.0
        self.positions = {} 
        self.fills = []

    def update_tier_data(self, taker: float, maker: float, level: str = "VIP 0"):
        """Updates the internal rates and tier label based on live OKX data."""
        self.taker_rate = abs(float(taker))
        self.maker_rate = abs(float(maker))
        self.tier_label = str(level).upper()

    def calculate_preflight_metrics(self, price: float, size: float, tp_price: float = None, sl_price: float = None) -> Dict[str, Any]:
        """
        Calculates tactical metrics with Liquidity-Aware logic:
        - Entry: Taker Rate (Conservative)
        - TP Exit: Maker Rate (Resting Limit)
        - SL Exit: Taker Rate (Market/Emergency)
        """
        if price <= 0 or size <= 0:
            return {"fee": 0.0, "break_even": 0.0, "net_tp": 0.0, "net_sl": 0.0}

        entry_fee = price * size * self.taker_rate
        
        # HURDLE: (Entry * (1 + Taker)) / (1 - Maker)
        # Assumes a successful trade clears via a resting TP (Maker).
        break_even = (price * (1 + self.taker_rate)) / (1 - self.maker_rate)
        
        total_est_friction = entry_fee
        net_tp = 0.0
        net_sl = 0.0
        
        if tp_price:
            # TP uses Maker Rate
            tp_exit_fee = tp_price * size * self.maker_rate
            net_tp = ((tp_price - price) * size) - (entry_fee + tp_exit_fee)
            total_est_friction = entry_fee + tp_exit_fee

        if sl_price:
            # SL uses Taker Rate (Emergency)
            sl_exit_fee = sl_price * size * self.taker_rate
            net_sl = ((sl_price - price) * size) - (entry_fee + sl_exit_fee)
            # If no TP set, show friction for the SL side
            if not tp_price:
                total_est_friction = entry_fee + sl_exit_fee

        return {
            "fee": total_est_friction,
            "break_even": break_even,
            "net_tp": net_tp,
            "net_sl": net_sl
        }

    def record_confirmed_fill(self, inst_id: str, side: str, price: float, size: float, tag: str = "Manual"):
        """Records a successful fill in the ledger."""
        # Use Taker Rate for the realized ledger as a conservative standard
        fee_usd = price * size * self.taker_rate
        self.total_fees_paid += fee_usd

        fill = {
            "time": time.strftime("%H:%M:%S"),
            "inst": inst_id,
            "side": side.upper(),
            "px": price,
            "sz": size,
            "fee": fee_usd,
            "tag": tag
        }
        self.fills.append(fill)

        # Basic position tracking for session PnL
        pos = self.positions.get(inst_id, {"size": 0.0, "avg_price": 0.0})
        curr_sz = pos["size"]
        curr_avg = pos["avg_price"]

        if side.upper() == "BUY":
            new_sz = curr_sz + size
            if curr_sz < 0:
                covered = min(abs(curr_sz), size)
                self.realized_pnl_gross += (curr_avg - price) * covered
            if new_sz > 0:
                pos["avg_price"] = ((curr_avg * max(0, curr_sz)) + (price * size)) / new_sz
            pos["size"] = new_sz
        elif side.upper() == "SELL":
            new_sz = curr_sz - size
            if curr_sz > 0:
                sold = min(curr_sz, size)
                self.realized_pnl_gross += (price - curr_avg) * sold
            if new_sz < 0:
                pos["avg_price"] = ((curr_avg * min(0, curr_sz)) + (price * -size)) / new_sz
            pos["size"] = new_sz

        self.positions[inst_id] = pos

    def get_session_summary(self, current_market_prices: Dict[str, float]) -> Dict[str, Any]:
        """Calculates the total NET SCORE of the session."""
        unrealized_gross = 0.0
        for inst, pos in self.positions.items():
            if pos["size"] == 0: continue
            mark_price = current_market_prices.get(inst, pos["avg_price"])
            unrealized_gross += (mark_price - pos["avg_price"]) * pos["size"]

        return {
            "net": (self.realized_pnl_gross + unrealized_gross) - self.total_fees_paid,
            "fees": self.total_fees_paid
        }

File: CPP/Accountant.cpp
#include "Accountant.hpp"
#include <cmath>
#include <algorithm>
#include <iomanip>
#include <sstream>

PnLAccountant::PnLAccountant(double taker_fee_rate, double maker_fee_rate)
    : taker_rate(std::abs(taker_fee_rate)),
      maker_rate(std::abs(maker_fee_rate)),
      tier_label("PENDING"),
      realized_pnl_gross(0.0),
      total_fees_paid(0.0) {}

void PnLAccountant::update_tier_data(double taker, double maker, const std::string& level) {
    taker_rate = std::abs(taker);
    maker_rate = std::abs(maker);
    tier_label = level;
    std::transform(tier_label.begin(), tier_label.end(), tier_label.begin(), ::toupper);
}

PreflightMetrics PnLAccountant::calculate_preflight_metrics(double price, double size, double tp_price, double sl_price) {
    if (price <= 0 || size <= 0) {
        return {0.0, 0.0, 0.0, 0.0};
    }

    double entry_fee = price * size * taker_rate;

    // HURDLE: (Entry * (1 + Taker)) / (1 - Maker)
    double break_even = (price * (1.0 + taker_rate)) / (1.0 - maker_rate);

    double total_est_friction = entry_fee;
    double net_tp = 0.0;
    double net_sl = 0.0;

    if (tp_price > 0) {
        double tp_exit_fee = tp_price * size * maker_rate;
        net_tp = ((tp_price - price) * size) - (entry_fee + tp_exit_fee);
        total_est_friction = entry_fee + tp_exit_fee;
    }

    if (sl_price > 0) {
        double sl_exit_fee = sl_price * size * taker_rate;
        net_sl = ((sl_price - price) * size) - (entry_fee + sl_exit_fee);
        if (tp_price <= 0) {
            total_est_friction = entry_fee + sl_exit_fee;
        }
    }

    return {total_est_friction, break_even, net_tp, net_sl};
}

void PnLAccountant::record_confirmed_fill(const std::string& inst_id, const std::string& side, double price, double size, const std::string& tag) {
    double fee_usd = price * size * taker_rate;
    total_fees_paid += fee_usd;

    std::string side_upper = side;
    std::transform(side_upper.begin(), side_upper.end(), side_upper.begin(), ::toupper);

    FillRecord fill = {
        get_current_time_str(),
        inst_id,
        side_upper,
        price,
        size,
        fee_usd,
        tag
    };
    fills.push_back(fill);

    Position& pos = positions[inst_id];
    double curr_sz = pos.size;
    double curr_avg = pos.avg_price;

    if (side_upper == "BUY") {
        double new_sz = curr_sz + size;
        if (curr_sz < 0) {
            double covered = std::min(std::abs(curr_sz), size);
            realized_pnl_gross += (curr_avg - price) * covered;
        }
        if (new_sz > 0) {
            pos.avg_price = ((curr_avg * std::max(0.0, curr_sz)) + (price * size)) / new_sz;
        }
        pos.size = new_sz;
    } else if (side_upper == "SELL") {
        double new_sz = curr_sz - size;
        if (curr_sz > 0) {
            double sold = std::min(curr_sz, size);
            realized_pnl_gross += (price - curr_avg) * sold;
        }
        if (new_sz < 0) {
            pos.avg_price = ((curr_avg * std::min(0.0, curr_sz)) + (price * -size)) / new_sz;
        }
        pos.size = new_sz;
    }
}

SessionSummary PnLAccountant::get_session_summary(const std::map<std::string, double>& current_market_prices) {
    double unrealized_gross = 0.0;
    for (auto const& [inst, pos] : positions) {
        if (pos.size == 0) continue;

        auto it = current_market_prices.find(inst);
        double mark_price = (it != current_market_prices.end()) ? it->second : pos.avg_price;
        unrealized_gross += (mark_price - pos.avg_price) * pos.size;
    }

    return { (realized_pnl_gross + unrealized_gross) - total_fees_paid, total_fees_paid };
}

std::string PnLAccountant::get_current_time_str() {
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%H:%M:%S");
    return oss.str();
}
